README.md
# Crux Docker

Project to start running [Crux] on docker in AWS ECS.

The Crux-Docker project provides a template with two key goals; a main function capable of running Crux in a
production environment and a container to host the Crux service.

## Pre-Conditions

This library assumes that you have an AWS account that follows best practices for infrastructure. That being said, you
are expected to have isolated CloudFormation stacks that each have dedicated purposes. 

The host AWS account must have access to a Kafka cluster as the Kafka cluster will act as the primary data store for
data indexed by Crux.

## Issues

Feel free to submit issues at any time! Please include steps to reproduce any issue you encounter.

You may also use issues to submit questions. Questions will result in more detailed documentation.

## Configuration

### AWS Configuration

#### Security Groups

The Crux service expects that a Kafka service is running. Either the AWS Managed Service for Kafka (MSK) or a manual
service of Kafka can be used as the data store for Crux. In order for the Crux service to connect to Kafka, you should
grant access to the `crux-host-<stack-name>` security group on ports `2181` and `9092`.

Access to the Crux service is restricted using the `crux-client-<stack-name>`. Add the `crux-client-<stack-name>`
security group to any service that should be able to connect to Crux. By default, clients should connect to the Crux
service using port `3000`.   

### Environment Variables

| Name                        | Description                                     | Example           |
| --------------------------- | ----------------------------------------------- | ----------------- |
| HEALTH_CHECK_THRESHOLD      | Number of health check failures before shutdown | 3                 |
| HEALTH_CHECK_INTERVAL       | Frequency of health checks (sec)                | 5                 |
| HEALTH_CHECK_WAIT_TIME      | Wait time before starting health checks (sec)   | 5                 |
| STORAGE_DIR                 | Data storage directory                          | `/dev/crrux-data` |
| KAFKA_BOOTSTRAP_SERVER      | Kafka Bootstrap Server Address                  | `localhost:9092`  |
| SERVER_PORT                 | Crux HTTP API Server Port                       | 3000              |

## License

The MIT License (MIT)

Copyright © 2019-2020 callahanwilliam.com

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated 
documentation files (the "Software"), to deal in the Software without restriction, including without limitation the 
rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit 
persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the 
Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE 
WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR  
COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR 
OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

[Crux]:https://github.com/juxt/crux

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [0.1.2-alpha] - 2020-02-16
### Changed
- Upgraded to crux version `19.12-1.6.0-alpha`
- Updated clojure documentation
- Removed unused environment variables

## [0.1.1-alpha] - 2020-02-13
### Changed
- Added production logging configuration
- Added local health checks

## [0.1.0-alpha] - 2020-02-13
### Changed
- cloudformation.yml - Added DeploymentConfiguration
- LICENSE - Changed to MIT
- Added comments to CodeBuild Approval dialog

### Removed
- Unused `LOG_DIR` environment variable from Dockerfile

[0.1.2-alpha]: https://github.com/WillCallahan/crux-docker/compare/0.1.1-alpha...HEAD
[0.1.1-alpha]: https://github.com/WillCallahan/crux-docker/compare/0.1.1-alpha...0.1.2-alpha
[0.1.0-alpha]: https://github.com/WillCallahan/crux-docker/compare/0.1.0-alpha...0.1.1-alpha


test/crux_docker/core_test.clj
(ns crux-docker.core-test
  (:require [clojure.test :refer :all]
            [crux-docker.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


project.clj
(def crux-version (or (System/getenv "CRUX_VERSION") "19.12-1.6.0-alpha"))
(defproject crux-docker "0.1.0-SNAPSHOT"
  :description "FIXME: write description"
  :url "http://example.com/FIXME"
  :license {:name "EPL-2.0 OR GPL-2.0-or-later WITH Classpath-exception-2.0"
            :url "https://www.eclipse.org/legal/epl-2.0/"}
  :repositories [["releases" {:url "https://repo.clojars.org" :creds :gpg}]]
  :dependencies [[org.clojure/clojure "1.10.0"]
                 [org.apache.logging.log4j/log4j-api "2.11.0"]
                 [org.apache.logging.log4j/log4j-core "2.11.0"]
                 [org.apache.logging.log4j/log4j-1.2-api "2.11.0"]
                 [org.apache.logging.log4j/log4j-slf4j-impl "2.11.0"]
                 [org.clojure/tools.logging "0.4.1"]
                 [juxt/crux-core ~crux-version]
                 [juxt/crux-rocksdb ~crux-version]
                 [juxt/crux-kafka ~crux-version]
                 [juxt/crux-kafka-connect ~crux-version]
                 [juxt/crux-http-server ~crux-version]
                 [juxt/crux-rdf ~crux-version]
                 [juxt/crux-core ~crux-version]
                 [juxt/crux-http-client ~crux-version]]
  :repl-options {:init-ns crux-docker.core}
  :jar-name "crux-docker.jar"
  :uberjar-name "crux-docker-standalone.jar"
  :main crux-docker.core
  :aot [crux-docker.core])


src/crux_docker/core.clj
(ns crux-docker.core
  "Crux indexing server with local health checks"
  (:require [clojure.tools.logging :as log]
            [crux.node :as n]
            [crux.api :as crux]
            [crux.http-server :as srv]
            [crux.io :as cio])
  (:import [crux.api ICruxAPI])
  (:gen-class))

(def storage-dir "crux-storage")

(defn get-env
  [key default parse]
  (if-let [env (System/getenv key)]
    (parse env)
    default))

(defn parse-int
  [i]
  (Integer/parseInt i))

(defn dev-node-option-defaults []
  {:crux.node/topology                    'crux.kafka/topology
   :crux.node/kv-store                    'crux.kv.rocksdb/kv
   :crux.kv/db-dir                        (get-env "STORAGE_DIR" (str storage-dir "/data") str)
   :crux.kv.memdb/persist-on-close?       true
   :crux.kafka/bootstrap-servers          (get-env "KAFKA_BOOTSTRAP_SERVER" "10.0.127.51:9092" str)})

(defn dev-http-option-defaults []
  {:server-port (get-env "SERVER_PORT" 3000 parse-int)
   :cors-access-control
                [:access-control-allow-origin [#".*"]
                 :access-control-allow-headers ["X-Requested-With"
                                                "Content-Type"
                                                "Cache-Control"
                                                "Origin"
                                                "Accept"
                                                "Authorization"
                                                "X-Custom-Header"]
                 :access-control-allow-methods [:get :post]]})

(def node-options (dev-node-option-defaults))

(def http-options (dev-http-option-defaults))

(def ^ICruxAPI node)

(defn start-dev-node ^crux.api.ICruxAPI [node-options http-options]
  (let [started (atom [])]
    (try
      (let [cluster-node (doto (crux/start-node node-options)
                           (->> (swap! started conj)))
            http-server (srv/start-http-server cluster-node http-options)]
        (assoc cluster-node
          :http-server http-server))
      (catch Throwable t
        (doseq [c (reverse @started)]
          (cio/try-close c))
        (throw t)))))

(defn stop-dev-node ^crux.api.ICruxAPI [{:keys [http-server embedded-kafka] :as node}]
  (doseq [c [http-server node embedded-kafka]]
    (cio/try-close c)))

(defn start []
  (alter-var-root #'node (fn [_] (start-dev-node node-options http-options)))
  :started)

(defn stop []
  (when (and (bound? #'node)
             (not (nil? node)))
    (alter-var-root #'node stop-dev-node))
  :stopped)

(defn delete-storage []
  (stop)
  (cio/delete-dir storage-dir)
  :ok)

(n/install-uncaught-exception-handler!)

(defn- health-check-internal
  [node]
  (log/info "Getting status")
  (try {:healthy (crux/status node)}
       (catch Exception e {:unhealthy e})))

(defn health-check
  [node]
  (if node
    (let [status (health-check-internal node)]
      (log/infof "Got Status: %s" status)
      status)
    (do
      (log/info "Node is not defied")
      :unhealthy (IllegalStateException. "Node is not defined"))))

(def health-check-threshold
  (get-env "HEALTH_CHECK_THRESHOLD" 3 parse-int))

(def health-check-interval
  (* (get-env "HEALTH_CHECK_INTERVAL" 5 parse-int) 1000))

(def health-check-wait-time
  (* (get-env "HEALTH_CHECK_WAIT_TIME" 5 parse-int) 1000))

(def heal-check-failures (atom 0))

(defn -main
  [& args]
  (log/info "Starting crux node")
  (start)
  (log/infof "Waiting for node to start (%s sec)" health-check-wait-time)
  (Thread/sleep health-check-wait-time)
  (log/infof "Starting health checks  Threshold=%s Interval=%s" health-check-threshold health-check-interval)
  (while (<= @heal-check-failures health-check-threshold)
    (do
      (if (contains? (health-check node) :healthy)
        (swap! heal-check-failures #(max 0 (dec %)))
        (swap! heal-check-failures inc))
      (log/infof "Health check failures %s" @heal-check-failures)
      (Thread/sleep health-check-interval)))
  (log/fatal "Shutting down; health checks failed")
  (stop)
  (System/exit 1))